File: token-flamegraph/flamegraph.py
"""
Convert a parsed Session into visualization-ready data structures.

Produces:
- Per-turn waterfall data (output token breakdown)
- Aggregate category breakdown
- Context growth over turns
- Tool call detail list
"""

import json
from dataclasses import dataclass, field
from parser import Session, Turn, ToolCall


@dataclass
class ToolSpan:
    """A tool call rendered as a span in the waterfall."""
    name: str
    tokens: int  # output tokens (input + output of the tool)
    category: str  # "read", "edit", "bash", "search", "agent", "other"
    children: list["ToolSpan"] = field(default_factory=list)
    preview: str = ""

    def total_tokens(self) -> int:
        return self.tokens + sum(c.total_tokens() for c in self.children)


@dataclass
class TurnData:
    """Visualization data for one assistant turn."""
    index: int
    thinking_tokens: int
    text_tokens: int
    tool_spans: list[ToolSpan]
    input_tokens: int
    cache_read_tokens: int
    fresh_input_tokens: int
    # Duration from teleport result events (ms)
    duration_ms: int = 0
    duration_api_ms: int = 0

    @property
    def output_tokens(self) -> int:
        """Total output-side tokens (what the agent chose to produce)."""
        return (
            self.thinking_tokens
            + self.text_tokens
            + sum(ts.total_tokens() for ts in self.tool_spans)
        )

    @property
    def tool_tokens(self) -> int:
        return sum(ts.total_tokens() for ts in self.tool_spans)

    @property
    def efficiency_score(self) -> float:
        """0-100 score: how efficiently this turn used tokens.

        Penalizes:
        - Write-heavy turns (Write is wasteful vs Edit)
        - High thinking-to-action ratio
        - Large output with few tools (verbose)
        - Low cache hit rate

        Rewards:
        - Edit usage over Write
        - High tool-to-output ratio (concise)
        - Good cache utilization
        """
        out = self.output_tokens
        if out == 0:
            return 0.0

        score = 100.0
        num_tools = len(self.tool_spans)

        # Penalty: insufficient data — short turns with <500 tokens get
        # capped at 80 since we can't meaningfully assess efficiency
        if out < 500:
            score = min(score, 80.0)

        # Penalty: Write-heavy (each Write token costs more than Edit)
        write_tokens = sum(
            ts.total_tokens() for ts in self.tool_spans
            if ts.name.startswith("Write")
        )
        if write_tokens > 0:
            write_ratio = write_tokens / out
            score -= write_ratio * 40

        # Penalty: high thinking ratio (>30% thinking = overthinking)
        think_ratio = self.thinking_tokens / out
        if think_ratio > 0.3:
            score -= (think_ratio - 0.3) * 50

        # Penalty: verbose output (>5K tokens with few tools = wall of text)
        if out > 5000 and num_tools < 3:
            score -= 20

        # Penalty: no tool calls at all (pure text generation, rarely efficient)
        if num_tools == 0 and out > 200:
            score -= 15

        # Penalty: low cache rate (under 80%)
        if self.input_tokens > 0:
            cache_rate = self.cache_read_tokens / self.input_tokens
            if cache_rate < 0.8:
                score -= (0.8 - cache_rate) * 30

        # Bonus: Edit usage (efficient modification)
        edit_tokens = sum(
            ts.total_tokens() for ts in self.tool_spans
            if ts.name.startswith("Edit")
        )
        if edit_tokens > 0 and out > 0:
            score += min(10, (edit_tokens / out) * 20)

        # Bonus: high tool density (concise, action-oriented)
        if out > 0 and num_tools > 0:
            density = num_tools / (out / 1000)
            score += min(10, density * 2)

        return round(max(0, min(100, score)), 1)


@dataclass
class SessionViz:
    """All visualization data for a session."""
    turns: list[TurnData]
    model: str
    session_id: str

    # Aggregates
    total_thinking: int = 0
    total_text: int = 0
    total_tool: int = 0
    total_input: int = 0
    total_cache: int = 0
    total_output: int = 0
    num_tool_calls: int = 0

    # Duration totals (from teleport result events)
    total_duration_ms: int = 0
    total_api_duration_ms: int = 0

    # Per-category tool breakdown
    tool_category_tokens: dict[str, int] = field(default_factory=dict)

    # Top tools by token usage
    top_tools: list[tuple[str, int]] = field(default_factory=list)


def session_to_viz(session: Session) -> SessionViz:
    """Convert a parsed session into visualization data."""
    turns = []
    tool_totals: dict[str, int] = {}
    category_totals: dict[str, int] = {}
    num_tools = 0

    for turn in session.turns:
        if turn.role != "assistant":
            continue

        tool_spans = [_tool_to_span(tc) for tc in turn.tool_calls]
        tool_tok = sum(ts.total_tokens() for ts in tool_spans)
        # Text tokens = total output minus thinking minus tool input estimates
        text_tok = max(0, turn.output_tokens - turn.thinking_tokens - sum(
            tc.input_tokens for tc in turn.tool_calls
        ))

        td = TurnData(
            index=turn.index,
            thinking_tokens=turn.thinking_tokens,
            text_tokens=text_tok,
            tool_spans=tool_spans,
            input_tokens=turn.input_tokens,
            cache_read_tokens=turn.cache_read_tokens,
            fresh_input_tokens=turn.input_tokens - turn.cache_read_tokens,
            duration_ms=turn.duration_ms,
            duration_api_ms=turn.duration_api_ms,
        )
        turns.append(td)

        # Accumulate tool stats
        for ts in tool_spans:
            _accumulate_tool(ts, tool_totals, category_totals)
            num_tools += 1
            num_tools += _count_children(ts)

    viz = SessionViz(
        turns=turns,
        model=session.model,
        session_id=session.session_id,
        total_thinking=sum(t.thinking_tokens for t in turns),
        total_text=sum(t.text_tokens for t in turns),
        total_tool=sum(t.tool_tokens for t in turns),
        total_input=sum(t.input_tokens for t in turns),
        total_cache=sum(t.cache_read_tokens for t in turns),
        total_output=sum(t.output_tokens for t in turns),
        total_duration_ms=sum(t.duration_ms for t in turns),
        total_api_duration_ms=sum(t.duration_api_ms for t in turns),
        num_tool_calls=num_tools,
        tool_category_tokens=category_totals,
        top_tools=sorted(tool_totals.items(), key=lambda x: -x[1])[:15],
    )
    return viz


def _tool_to_span(tc: ToolCall) -> ToolSpan:
    name = _tool_label(tc)
    cat = _tool_category(tc.name)
    children = [_tool_to_span(c) for c in tc.children]
    return ToolSpan(
        name=name,
        tokens=tc.input_tokens + tc.output_tokens,
        category=cat,
        children=children,
        preview=tc.input_preview[:100],
    )


def _tool_category(name: str) -> str:
    categories = {
        "Read": "read", "Grep": "search", "Glob": "search",
        "Edit": "write", "Write": "write",
        "Bash": "bash", "Agent": "agent",
    }
    return categories.get(name, "other")


def _tool_label(tc: ToolCall) -> str:
    name = tc.name
    preview = tc.input_preview
    if name in ("Read", "Edit", "Write") and "file_path" in preview:
        try:
            data = json.loads(preview)
            fp = data.get("file_path", "")
            short = fp.split("/")[-1] if "/" in fp else fp
            return f"{name}({short})"
        except (json.JSONDecodeError, TypeError):
            pass
    if name == "Bash" and "command" in preview:
        try:
            data = json.loads(preview)
            cmd = data.get("command", "")[:35]
            return f"Bash({cmd})"
        except (json.JSONDecodeError, TypeError):
            pass
    if name == "Agent":
        desc = preview.replace("[Agent] ", "")[:50]
        return f"Agent({desc})"
    return name


def _accumulate_tool(ts: ToolSpan, tool_totals: dict, cat_totals: dict):
    tool_totals[ts.name] = tool_totals.get(ts.name, 0) + ts.tokens
    cat_totals[ts.category] = cat_totals.get(ts.category, 0) + ts.tokens
    for child in ts.children:
        _accumulate_tool(child, tool_totals, cat_totals)


def _count_children(ts: ToolSpan) -> int:
    return sum(1 + _count_children(c) for c in ts.children)


def viz_to_json(viz: SessionViz) -> dict:
    """Serialize visualization data for embedding in HTML."""
    return {
        "turns": [
            {
                "index": t.index,
                "thinking": t.thinking_tokens,
                "text": t.text_tokens,
                "tools": [_span_to_dict(s) for s in t.tool_spans],
                "toolTokens": t.tool_tokens,
                "outputTokens": t.output_tokens,
                "inputTokens": t.input_tokens,
                "cacheRead": t.cache_read_tokens,
                "freshInput": t.fresh_input_tokens,
                "durationMs": t.duration_ms,
                "durationApiMs": t.duration_api_ms,
            }
            for t in viz.turns
        ],
        "totals": {
            "thinking": viz.total_thinking,
            "text": viz.total_text,
            "tool": viz.total_tool,
            "input": viz.total_input,
            "cache": viz.total_cache,
            "output": viz.total_output,
            "toolCalls": viz.num_tool_calls,
            "durationMs": viz.total_duration_ms,
            "durationApiMs": viz.total_api_duration_ms,
        },
        "categoryTokens": viz.tool_category_tokens,
        "topTools": [{"name": n, "tokens": t} for n, t in viz.top_tools],
        "model": viz.model,
    }


def _span_to_dict(ts: ToolSpan) -> dict:
    return {
        "name": ts.name,
        "tokens": ts.tokens,
        "totalTokens": ts.total_tokens(),
        "category": ts.category,
        "children": [_span_to_dict(c) for c in ts.children],
    }

File: token-flamegraph-rs/src/flamegraph.rs
use crate::parser::{Session, ToolCall};
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct ToolSpan {
    pub name: String,
    pub tokens: usize,
    pub category: String,
    pub children: Vec<ToolSpan>,
    #[serde(rename = "totalTokens")]
    pub total_tokens: usize,
}

impl ToolSpan {
    pub fn compute_total(&mut self) -> usize {
        let child_total: usize = self.children.iter_mut().map(|c| c.compute_total()).sum();
        self.total_tokens = self.tokens + child_total;
        self.total_tokens
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct TurnViz {
    pub index: usize,
    pub thinking: usize,
    pub text: usize,
    pub tools: Vec<ToolSpan>,
    #[serde(rename = "toolTokens")]
    pub tool_tokens: usize,
    #[serde(rename = "outputTokens")]
    pub output_tokens: usize,
    #[serde(rename = "inputTokens")]
    pub input_tokens: usize,
    #[serde(rename = "cacheRead")]
    pub cache_read: usize,
    #[serde(rename = "freshInput")]
    pub fresh_input: usize,
    #[serde(rename = "durationMs")]
    pub duration_ms: u64,
    #[serde(rename = "durationApiMs")]
    pub duration_api_ms: u64,
    #[serde(skip_serializing)]
    pub efficiency_score: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct VizData {
    pub turns: Vec<TurnViz>,
    pub totals: Totals,
    #[serde(rename = "categoryTokens")]
    pub category_tokens: std::collections::HashMap<String, usize>,
    #[serde(rename = "topTools")]
    pub top_tools: Vec<TopTool>,
    pub model: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct Totals {
    pub thinking: usize,
    pub text: usize,
    pub tool: usize,
    pub input: usize,
    pub cache: usize,
    pub output: usize,
    #[serde(rename = "toolCalls")]
    pub tool_calls: usize,
    #[serde(rename = "durationMs")]
    pub duration_ms: u64,
    #[serde(rename = "durationApiMs")]
    pub duration_api_ms: u64,
}

#[derive(Debug, Clone, Serialize)]
pub struct TopTool {
    pub name: String,
    pub tokens: usize,
}

fn tool_category(name: &str) -> &'static str {
    if name.starts_with("Read") { "read" }
    else if name.starts_with("Edit") || name.starts_with("Write") { "write" }
    else if name.starts_with("Bash") { "bash" }
    else if name.starts_with("Grep") || name.starts_with("Glob") { "search" }
    else if name.starts_with("Agent") { "agent" }
    else { "other" }
}

fn tool_to_span(tc: &ToolCall) -> ToolSpan {
    let cat = tool_category(&tc.name).to_string();
    let children: Vec<ToolSpan> = tc.children.iter().map(tool_to_span).collect();
    let mut span = ToolSpan {
        name: tc.name.clone(),
        tokens: tc.input_tokens + tc.output_tokens,
        category: cat,
        children,
        total_tokens: 0,
    };
    span.compute_total();
    span
}

fn efficiency_score(turn: &TurnViz) -> f64 {
    let out = turn.output_tokens;
    if out == 0 { return 0.0; }
    let mut score = 100.0f64;
    let num_tools = turn.tools.len();

    if out < 500 { score = score.min(80.0); }

    let write_tokens: usize = turn.tools.iter()
        .filter(|t| t.name.starts_with("Write"))
        .map(|t| t.total_tokens).sum();
    if write_tokens > 0 {
        score -= (write_tokens as f64 / out as f64) * 40.0;
    }

    let think_ratio = turn.thinking as f64 / out as f64;
    if think_ratio > 0.3 {
        score -= (think_ratio - 0.3) * 50.0;
    }

    if out > 5000 && num_tools < 3 { score -= 20.0; }
    if num_tools == 0 && out > 200 { score -= 15.0; }

    if turn.input_tokens > 0 {
        let cache_rate = turn.cache_read as f64 / turn.input_tokens as f64;
        if cache_rate < 0.8 {
            score -= (0.8 - cache_rate) * 30.0;
        }
    }

    let edit_tokens: usize = turn.tools.iter()
        .filter(|t| t.name.starts_with("Edit"))
        .map(|t| t.total_tokens).sum();
    if edit_tokens > 0 && out > 0 {
        score += (edit_tokens as f64 / out as f64 * 20.0).min(10.0);
    }

    if out > 0 && num_tools > 0 {
        let density = num_tools as f64 / (out as f64 / 1000.0);
        score += (density * 2.0).min(10.0);
    }

    score.max(0.0).min(100.0)
}

pub fn session_to_viz(session: &Session) -> VizData {
    let mut turns = Vec::new();
    let mut tool_totals: std::collections::HashMap<String, usize> = std::collections::HashMap::new();
    let mut category_totals: std::collections::HashMap<String, usize> = std::collections::HashMap::new();
    let mut num_tool_calls = 0usize;

    for turn in &session.turns {
        if turn.role != "assistant" { continue; }

        let tool_spans: Vec<ToolSpan> = turn.tool_calls.iter().map(tool_to_span).collect();
        let tool_tok: usize = tool_spans.iter().map(|s| s.total_tokens).sum();
        let tool_input_est: usize = turn.tool_calls.iter().map(|tc| tc.input_tokens).sum();
        let text_tok = turn.output_tokens.saturating_sub(turn.thinking_tokens).saturating_sub(tool_input_est);

        let mut tv = TurnViz {
            index: turn.index,
            thinking: turn.thinking_tokens,
            text: text_tok,
            tool_tokens: tool_tok,
            output_tokens: turn.thinking_tokens + text_tok + tool_tok,
            input_tokens: turn.input_tokens,
            cache_read: turn.cache_read_tokens,
            fresh_input: turn.input_tokens.saturating_sub(turn.cache_read_tokens),
            duration_ms: turn.duration_ms,
            duration_api_ms: turn.duration_api_ms,
            tools: tool_spans,
            efficiency_score: 0.0,
        };
        tv.efficiency_score = efficiency_score(&tv);
        turns.push(tv);

        for tc in &turn.tool_calls {
            *tool_totals.entry(tc.name.clone()).or_insert(0) += tc.input_tokens + tc.output_tokens;
            *category_totals.entry(tool_category(&tc.name).to_string()).or_insert(0) += tc.input_tokens + tc.output_tokens;
            num_tool_calls += 1;
            num_tool_calls += tc.children.len();
        }
    }

    let mut top: Vec<TopTool> = tool_totals.into_iter()
        .map(|(name, tokens)| TopTool { name, tokens })
        .collect();
    top.sort_by(|a, b| b.tokens.cmp(&a.tokens));
    top.truncate(15);

    let totals = Totals {
        thinking: turns.iter().map(|t| t.thinking).sum(),
        text: turns.iter().map(|t| t.text).sum(),
        tool: turns.iter().map(|t| t.tool_tokens).sum(),
        input: turns.iter().map(|t| t.input_tokens).sum(),
        cache: turns.iter().map(|t| t.cache_read).sum(),
        output: turns.iter().map(|t| t.output_tokens).sum(),
        tool_calls: num_tool_calls,
        duration_ms: turns.iter().map(|t| t.duration_ms).sum(),
        duration_api_ms: turns.iter().map(|t| t.duration_api_ms).sum(),
    };

    VizData { turns, totals, category_tokens: category_totals, top_tools: top, model: session.model.clone() }
}
